File: mtui/mcp/_argv.py
"""Reassemble argparse argv from a tool-call kwargs dict.

The MCP server delivers a tool call as a dict of ``{dest: value}`` keyed by the
synthesised parameter names from :mod:`mtui.mcp._schema`. To run the
underlying :class:`mtui.commands._command.Command` we must turn that
dict back into a token list ``argparse`` can parse. The mapping is the
inverse of :func:`mtui.mcp._schema.action_to_parameter`:

* ``store_true`` / ``store_false`` / ``store_const`` → emit the long
  flag iff the value is the "on" side.
* ``append`` → emit ``[flag, item]`` per element of the list value, EXCEPT
  when the action's ``nargs`` consumes a remainder/multi
  (``REMAINDER``/``*``/``+``/integer ``N`` — e.g. ``commit -m``, ``lock -c``):
  there the flag is emitted once followed by every token, so argparse rebuilds
  ``[[v1, v2, ...]]`` instead of swallowing a repeated flag as a value.
* Positional ``nargs=REMAINDER``/``*``/``+`` → append elements verbatim
  at the end (after every flag-shaped argument), preserving order.
* Optional ``nargs=REMAINDER`` (e.g. ``reject --message``) → emit the
  flag once followed by every token, but **into the positional tail** so
  it lands after every other flag. A REMAINDER optional consumes all
  remaining argv tokens, so emitting a later ``--flag`` behind it would
  be swallowed as a value; deferring it to the tail keeps it safe
  regardless of ``parser._actions`` declaration order.
* Optional scalar / other multi-value (``+``/``*``/N) → emit
  ``[long_flag, v1, v2, ...]`` (or ``[long_flag, str(value)]`` for a
  scalar) among the flags.
* Positional scalar → append ``str(value)`` after the flags.

The function returns a list of strings; callers (notably
:meth:`McpSession.run_command`) feed it to :func:`shlex.join` and then
to the per-command argparser.
"""

from __future__ import annotations

import argparse
from logging import getLogger
from typing import Any

logger = getLogger("mtui.mcp.argv")

#: Sentinel used by argparse for "no default supplied". We treat
#: kwargs equal to this as "client did not supply anything".
_NO_DEFAULT = object()


def _action_index(
    parser: argparse.ArgumentParser,
) -> dict[str, argparse.Action]:
    """Build a ``{dest: action}`` map for ``parser``.

    Multiple actions can share a ``dest`` (mutually exclusive groups);
    :mod:`mtui.mcp._schema` keeps only the first one, so we mirror that
    here. Used by :func:`kwargs_to_argv` to look up how each kwarg
    should be re-emitted.
    """
    index: dict[str, argparse.Action] = {}
    for action in parser._actions:  # noqa: SLF001 - argparse exposes only this
        if isinstance(action, argparse._SubParsersAction):
            continue
        if action.dest == argparse.SUPPRESS:
            continue
        index.setdefault(action.dest, action)
    return index


def _long_flag(action: argparse.Action) -> str:
    """Return the long ``--flag`` form, falling back to whatever exists.

    Every mtui optional that has a short form also has a long form, but
    we tolerate optional-only short forms defensively.
    """
    for opt in action.option_strings:
        if opt.startswith("--"):
            return opt
    return action.option_strings[0]


def kwargs_to_argv(
    parser: argparse.ArgumentParser,
    kwargs: dict[str, Any],
) -> list[str]:
    """Re-encode a tool-call kwargs dict as argparse-compatible argv.

    Arguments are emitted in the same order ``parser._actions`` defines
    them so the output is deterministic and easy to read in logs. All
    plain optional flags come out first, then the positional tail —
    argparse accepts positionals after flags but not interleaved with
    them when ``nargs=REMAINDER`` is involved. A REMAINDER-consuming
    *optional* (e.g. ``reject --message``) is also deferred to the tail
    so it is emitted after every other flag and cannot swallow one.

    Args:
        parser: The command's :class:`argparse.ArgumentParser`. Used to
            look up the action class behind each kwarg.
        kwargs: ``{dest: value}`` as delivered by the MCP server.

    Returns:
        A list of argv tokens ready to be passed through
        :func:`shlex.join` and back into :meth:`Command.parse_args`.

    """
    index = _action_index(parser)
    flag_tokens: list[str] = []
    positional_tail: list[str] = []

    # ---- synthetic-name routing (load_template -a / -k case) -----------
    # mtui.mcp._schema collapses mutex StoreAction groups that share a
    # dest into one parameter per long flag and attaches the mapping
    # here. Each non-None value is emitted via its underlying action's
    # long flag; the underlying actions are then skipped in the main
    # loop because nothing in ``kwargs`` references their real dest.
    synthetic_map: dict[str, argparse.Action] = getattr(
        parser,
        "_mtui_synthetic_dests",
        {},
    )
    synthetic_action_ids: set[int] = {id(a) for a in synthetic_map.values()}
    for syn_name, syn_action in synthetic_map.items():
        if syn_name not in kwargs:
            continue
        syn_value = kwargs[syn_name]
        if syn_value is None:
            continue
        flag_tokens.extend([_long_flag(syn_action), str(syn_value)])

    # ---- StoreConst enum routing (set_repo -A / -R case) ----------------
    # mtui.mcp._schema also collapses mutex StoreConstAction groups that
    # share a dest into one Literal parameter. The kwarg value is now
    # the chosen const, not a bool, so we walk the parser's mutex
    # groups and emit the long flag of the matching member action.
    # Action ids touched here are skipped by the main loop so the
    # legacy truthy-check below doesn't re-emit a stale flag.
    const_action_ids: set[int] = set()
    for group in parser._mutually_exclusive_groups:  # noqa: SLF001
        members = list(group._group_actions)  # noqa: SLF001
        if len(members) < 2:
            continue
        if not all(
            isinstance(a, argparse._StoreConstAction)
            for a in members  # noqa: SLF001
        ):
            continue
        dests = {a.dest for a in members}
        if len(dests) != 1:
            continue
        shared_dest = members[0].dest
        for a in members:
            const_action_ids.add(id(a))
        if shared_dest not in kwargs:
            continue
        chosen = kwargs[shared_dest]
        if chosen is None:
            continue
        for a in members:
            if a.const == chosen:
                flag_tokens.append(_long_flag(a))
                break

    for action in parser._actions:  # noqa: SLF001 - argparse exposes only this
        if id(action) in synthetic_action_ids or id(action) in const_action_ids:
            continue  # already handled by a synthetic pass above
        if getattr(action, "_mtui_mcp_hidden", False):
            # Never emit a REPL-only hidden flag (e.g. approve/reject --force):
            # build_parameters keeps it out of the MCP schema, and this makes the
            # encoder refuse to render it even if an extra kwarg slips past the
            # SDK's schema validation — the security property no longer depends on
            # Pydantic's default extra='ignore'.
            continue
        if action.dest not in index or index[action.dest] is not action:
            continue  # secondary action sharing a dest (e.g. load_template -k)
        if action.dest not in kwargs:
            continue
        value = kwargs[action.dest]

        # ---- boolean-shaped flags ------------------------------------
        if isinstance(action, argparse._StoreTrueAction):
            if value:
                flag_tokens.append(_long_flag(action))
            continue
        if isinstance(action, argparse._StoreFalseAction):
            if not value:
                flag_tokens.append(_long_flag(action))
            continue
        if isinstance(action, argparse._StoreConstAction):
            if value:
                flag_tokens.append(_long_flag(action))
            continue

        # ---- append ---------------------------------------------------
        if isinstance(action, argparse._AppendAction):
            if not value:
                continue
            flag = _long_flag(action)
            if action.nargs in (argparse.REMAINDER, "*", "+") or isinstance(
                action.nargs, int
            ):
                # append + a remainder/multi nargs (commit -m, lock -c) stores
                # ONE sub-list per flag occurrence and the command reads
                # value[0]. Emit the flag once followed by every token so
                # argparse rebuilds ``[[v1, v2, ...]]``; ``--flag v1 --flag v2``
                # would let REMAINDER swallow the second ``--flag`` as a value,
                # producing ``[[v1, '--flag', v2]]`` and a corrupted message.
                flag_tokens.append(flag)
                flag_tokens.extend(str(item) for item in value)
            else:
                for item in value:
                    flag_tokens.extend([flag, str(item)])
            continue

        # ---- positional ----------------------------------------------
        is_positional = not action.option_strings
        if is_positional:
            if value is None:
                continue
            if isinstance(value, list):
                positional_tail.extend(str(x) for x in value)
            else:
                positional_tail.append(str(value))
            continue

        # ---- optional scalar / multi-value flag ----------------------
        if value is None:
            continue
        flag = _long_flag(action)
        if isinstance(value, list):
            # ``nargs`` in ``+``, ``*``, integer N, or ``REMAINDER``
            # makes argparse consume the rest of the argv after the flag
            # token. We must emit ``[--flag, v1, v2, ...]`` once, not
            # ``[--flag, v1, --flag, v2]`` (the latter would be parsed
            # as a single value list ``[v1, '--flag', v2]``).
            if action.nargs == argparse.REMAINDER:
                # A REMAINDER optional swallows *every* remaining token,
                # including a later ``--flag``. Emitting it inside
                # ``flag_tokens`` is only safe when it happens to be the
                # last flag declared; route it into the positional tail so
                # it is always emitted after every other flag regardless of
                # ``parser._actions`` order (e.g. ``reject --message``).
                positional_tail.append(flag)
                positional_tail.extend(str(item) for item in value)
            else:
                flag_tokens.append(flag)
                flag_tokens.extend(str(item) for item in value)
        else:
            flag_tokens.extend([flag, str(value)])

    return flag_tokens + positional_tail

File: crates/mtui-mcp/src/argv.rs
//! Reconstruct a `clap` argv token list from a tool-call kwargs dict.
//!
//! The **inverse of [`crate::schema`]**: a tool call arrives as a
//! `{param_name: value}` object, and dispatching it through the *same* engine
//! the REPL uses ([`mtui_core::dispatch_argv`]) means turning that object back
//! into the argv `clap` re-parses. Done purely, introspecting the command's
//! built [`clap::Command`] — the identical `get_arguments()` surface
//! [`crate::schema`] reads.
//!
//! Deliberately absent: synthetic-dest routing for mutually-exclusive groups
//! (clap already gives each member a distinct arg id, so its kwarg maps to its
//! own long flag through the normal loop), and an "exactly one required"
//! pre-check (`clap::ArgGroup` enforces that on re-parse, as a clean
//! [`mtui_core::EngineError::Parse`]).
//!
//! # Ordering
//!
//! Plain optional flags first (in `get_arguments()` order), then the positional
//! tail. Append / multi-value flags route into the tail too: a greedy
//! (`num_args(1..)`) one consumes every token after it, so a later flag
//! (notably the base parser's `-T/--template`, declared after per-command args)
//! must not sit behind it. An `Append` flag is emitted once per element (the only
//! form clap parses when each occurrence takes one value); a non-`Append`
//! multi-value option once with all its values (see the loop below).

use clap::{Arg, ArgAction};
use serde_json::{Map, Value};

/// Re-encode a tool-call kwargs dict as `clap`-compatible argv.
///
/// `argv_prefix` is prepended verbatim — used by the MCP tool synthesiser's
/// subparser fan-out to inject a subcommand name (`["show"]` for the
/// `config_show` tool). Args absent from `kwargs` or whose value is JSON `null`
/// are skipped. Flags come first, then the positional tail (module docs for why).
#[must_use]
pub(crate) fn kwargs_to_argv(
    cmd: &clap::Command,
    kwargs: &Map<String, Value>,
    argv_prefix: &[String],
) -> Vec<String> {
    let mut flags: Vec<String> = Vec::new();
    let mut tail: Vec<String> = Vec::new();

    for arg in cmd.get_arguments() {
        let id = arg.get_id().as_str();
        if id == "help" || id == "version" {
            continue;
        }
        let Some(value) = kwargs.get(id) else {
            continue;
        };
        if value.is_null() {
            continue;
        }

        // Boolean-shaped: emit the long flag only on the flag's "on" side.
        match arg.get_action() {
            ArgAction::SetTrue => {
                if value.as_bool() == Some(true) {
                    flags.push(long_flag(arg));
                }
                continue;
            }
            ArgAction::SetFalse => {
                if value.as_bool() == Some(false) {
                    flags.push(long_flag(arg));
                }
                continue;
            }
            _ => {}
        }

        let items = as_items(value);
        if items.is_empty() {
            continue;
        }

        if is_positional(arg) {
            tail.extend(items);
            continue;
        }

        // Both shapes route into the tail so a later flag cannot be swallowed as
        // one of their values, but emit differently: `Append` repeats the flag
        // before *every* element (the only form a default `num_args = 1` Append
        // arg parses), while a non-`Append` multi-value option consumes many
        // values from one occurrence — repeating it would make the last
        // occurrence win and silently drop the earlier values.
        if is_multi(arg) {
            if matches!(arg.get_action(), ArgAction::Append) {
                for item in items {
                    tail.push(long_flag(arg));
                    tail.push(item);
                }
            } else {
                tail.push(long_flag(arg));
                tail.extend(items);
            }
            continue;
        }

        flags.push(long_flag(arg));
        flags.extend(items);
    }

    let mut out = Vec::with_capacity(argv_prefix.len() + flags.len() + tail.len());
    out.extend_from_slice(argv_prefix);
    out.append(&mut flags);
    out.append(&mut tail);
    out
}

/// The long `--flag` form. Every mtui optional with a short form also has a
/// long form, so the short/id fallbacks are purely defensive.
fn long_flag(arg: &Arg) -> String {
    arg.get_long()
        .map(|l| format!("--{l}"))
        .or_else(|| arg.get_short().map(|c| format!("-{c}")))
        .unwrap_or_else(|| arg.get_id().to_string())
}

/// Whether an arg is a positional.
fn is_positional(arg: &Arg) -> bool {
    arg.get_long().is_none() && arg.get_short().is_none()
}

/// Whether a value-taking arg accepts multiple tokens after one flag
/// (`ArgAction::Append`, or a `num_args` upper bound above one). Kept in sync
/// with [`crate::schema`]'s `is_list_arg`.
fn is_multi(arg: &Arg) -> bool {
    if matches!(arg.get_action(), ArgAction::Append) {
        return true;
    }
    arg.get_num_args().is_some_and(|r| r.max_values() > 1)
}

/// Flatten a JSON value into the argv tokens it contributes: arrays fan out per
/// element, scalars render to one token. A top-level `null` is filtered by the
/// caller; one nested inside an array is dropped here.
fn as_items(value: &Value) -> Vec<String> {
    match value {
        Value::Array(items) => items.iter().filter_map(render_scalar).collect(),
        other => render_scalar(other).into_iter().collect(),
    }
}

/// Render a single non-array JSON value into an argv token, or `None` for `null`.
fn render_scalar(value: &Value) -> Option<String> {
    match value {
        Value::Null => None,
        Value::String(s) => Some(s.clone()),
        Value::Bool(b) => Some(b.to_string()),
        Value::Number(n) => Some(n.to_string()),
        // Not a valid CLI scalar; stringify so the downstream parse error names
        // the offending value rather than panicking.
        other => Some(other.to_string()),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use mtui_core::{Registry, command_parser, register_all};
    use serde_json::json;

    /// Build a command's full parser (base template flags + its own `configure`)
    /// through the same `mtui-core` builder `dispatch_argv` re-parses, so these
    /// tests exercise exactly the parser real dispatch sees.
    fn parser_for(command: &str) -> clap::Command {
        let registry: Registry = register_all();
        let cmd = registry
            .get(command)
            .unwrap_or_else(|| panic!("command not registered: {command}"));
        command_parser(cmd.as_ref())
    }

    fn argv(command: &str, kwargs: Value) -> Vec<String> {
        let parser = parser_for(command);
        kwargs_to_argv(&parser, kwargs.as_object().unwrap(), &[])
    }

    /// The reconstructed argv must re-parse through the same parser.
    fn assert_reparses(command: &str, tokens: &[String]) {
        let parser = parser_for(command);
        parser
            .try_get_matches_from(tokens)
            .unwrap_or_else(|e| panic!("argv {tokens:?} failed to re-parse: {e}"));
    }

    // ------------------------------------------------------------- booleans

    #[test]
    fn store_true_on_emits_flag_off_omits() {
        // `openqa_overview --export` (SetTrue).
        let on = argv("openqa_overview", json!({ "export": true }));
        assert_eq!(on, vec!["--export"]);
        let off = argv("openqa_overview", json!({ "export": false }));
        assert!(off.is_empty(), "false SetTrue emits nothing: {off:?}");
        assert_reparses("openqa_overview", &on);
    }

    #[test]
    fn store_false_off_side_emits_flag() {
        // No mtui command uses SetFalse; pin the branch on a bespoke parser.
        let parser = clap::Command::new("probe").no_binary_name(true).arg(
            clap::Arg::new("color")
                .long("no-color")
                .action(ArgAction::SetFalse),
        );
        let off = kwargs_to_argv(&parser, json!({ "color": false }).as_object().unwrap(), &[]);
        assert_eq!(off, vec!["--no-color"]);
        let on = kwargs_to_argv(&parser, json!({ "color": true }).as_object().unwrap(), &[]);
        assert!(on.is_empty(), "true SetFalse emits nothing: {on:?}");
    }

    // -------------------------------------------------------------- scalars

    #[test]
    fn int_scalar_flag_renders_number_token() {
        // `openqa_overview --days` (u32 option).
        let out = argv("openqa_overview", json!({ "days": 5 }));
        assert_eq!(out, vec!["--days", "5"]);
        assert_reparses("openqa_overview", &out);
    }

    #[test]
    fn required_int_positional_lands_in_tail() {
        // `set_timeout timeout` (required u64 positional).
        let out = argv("set_timeout", json!({ "timeout": 30 }));
        assert_eq!(out, vec!["30"]);
        assert_reparses("set_timeout", &out);
    }

    #[test]
    fn enum_positional_emits_value() {
        // `set_log_level level` (PossibleValues positional).
        let out = argv("set_log_level", json!({ "level": "info" }));
        assert_eq!(out, vec!["info"]);
        assert_reparses("set_log_level", &out);
    }

    #[test]
    fn regenerate_rrid_maps_to_positional_not_template_selector() {
        // `{rrid: <RRID>}` must land on the `regenerate` positional, NOT the base
        // `-T/--template` loaded-template selector — otherwise the engine rejects
        // the unloaded RRID with `Template not loaded`.
        let out = argv("regenerate", json!({ "rrid": "SUSE:SLFO:1.2:6311" }));
        assert_eq!(out, vec!["SUSE:SLFO:1.2:6311"]);
        assert!(
            !out.iter().any(|t| t == "--template" || t == "-T"),
            "RRID must not route through the -T selector: {out:?}"
        );
        assert_reparses("regenerate", &out);
    }

    #[test]
    fn regenerate_kernel_hint_and_rrid_round_trip() {
        let out = argv(
            "regenerate",
            json!({ "kernel": true, "rrid": "SUSE:Maintenance:1:1" }),
        );
        // Boolean flag first, positional in the tail.
        assert_eq!(out, vec!["--kernel", "SUSE:Maintenance:1:1"]);
        assert_reparses("regenerate", &out);
    }

    // ---------------------------------------------------------------- lists

    #[test]
    fn append_single_value_flag_repeats_per_element() {
        // `approve -g/--group` (Append, default num_args = 1): clap rejects the
        // flag-once form `--group a b`, so the flag must repeat per element.
        // Same shape as `updates --review-group` / `-F`.
        let out = argv("approve", json!({ "group": ["qam-sle", "qam-teradata"] }));
        assert_eq!(out, vec!["--group", "qam-sle", "--group", "qam-teradata"]);
        assert_reparses("approve", &out);
    }

    #[test]
    fn run_hosts_and_command_round_trip_without_leaking_across_boundary() {
        // The silent-corruption case: `run`'s `-t/--target` (Append) and its
        // trailing `COMMAND` positional share the tail, so a flag-once
        // `--target h1 h2 uname -a` parses as hosts=["h1"], command=["h2",
        // "uname","-a"] — running `h2 uname -a` on h1. Repeating `--target` per
        // host keeps the boundary intact; no other test covers this interaction.
        let out = argv(
            "run",
            json!({ "hosts": ["h1", "h2"], "command": ["uname", "-a"] }),
        );
        assert_eq!(out, vec!["--target", "h1", "--target", "h2", "uname", "-a"]);
        assert_reparses("run", &out);
        let parsed = parser_for("run").try_get_matches_from(&out).unwrap();
        let hosts: Vec<&String> = parsed.get_many::<String>("hosts").unwrap().collect();
        assert_eq!(hosts, ["h1", "h2"]);
        let command: Vec<&String> = parsed.get_many::<String>("command").unwrap().collect();
        assert_eq!(command, ["uname", "-a"]);
    }

    #[test]
    fn updates_multi_group_and_fields_round_trip() {
        // The #415 surface: two review groups and two field names (one with an
        // embedded space) round-trip to the value lists the client sent.
        let out = argv(
            "updates",
            json!({
                "review_group": ["qam-sle", "qam-teradata"],
                "field": ["Rating", "Assigned Roles"],
            }),
        );
        assert_reparses("updates", &out);
        let parsed = parser_for("updates").try_get_matches_from(&out).unwrap();
        let groups: Vec<&String> = parsed.get_many::<String>("review_group").unwrap().collect();
        assert_eq!(groups, ["qam-sle", "qam-teradata"]);
        let fields: Vec<&String> = parsed.get_many::<String>("field").unwrap().collect();
        assert_eq!(fields, ["Rating", "Assigned Roles"]);
    }

    #[test]
    fn append_multi_flag_repeats_per_element_and_reparses_identically() {
        // `commit -m/--msg` (Append, num_args 1..): tail-routed so a trailing
        // base flag stays safe, and for `num_args(1..)` the repeated form parses
        // to the same value list as the flag-once one — pinned by re-parsing.
        let out = argv(
            "commit",
            json!({ "msg": ["hello", "world"], "template": "a:b:1:1" }),
        );
        // `-T` (flag) precedes the tail-routed `--msg`.
        assert_eq!(
            out,
            vec!["--template", "a:b:1:1", "--msg", "hello", "--msg", "world"]
        );
        assert_reparses("commit", &out);
        let parsed = parser_for("commit").try_get_matches_from(&out).unwrap();
        let msgs: Vec<&String> = parsed.get_many::<String>("msg").unwrap().collect();
        assert_eq!(msgs, ["hello", "world"]);
    }

    #[test]
    fn append_list_flag_with_choices_repeats_per_element() {
        // `openqa_overview --aggregated-groups` (Append + PossibleValues,
        // default num_args = 1): every element gets its own flag, else a
        // two-element call cannot re-parse.
        let parser = parser_for("openqa_overview");
        // Discover valid enum members from the parser so the re-parse succeeds.
        let arg = parser
            .get_arguments()
            .find(|a| a.get_id() == "aggregated_groups")
            .expect("aggregated_groups arg");
        let members: Vec<String> = arg
            .get_possible_values()
            .iter()
            .take(2)
            .map(|pv| pv.get_name().to_owned())
            .collect();
        assert!(!members.is_empty(), "aggregated_groups has choices");
        let out = argv(
            "openqa_overview",
            json!({ "aggregated_groups": members.clone() }),
        );
        // One `--aggregated-groups` immediately before each member, in order.
        let expected: Vec<String> = members
            .iter()
            .flat_map(|m| ["--aggregated-groups".to_owned(), m.clone()])
            .collect();
        assert_eq!(out, expected);
        assert_reparses("openqa_overview", &out);
    }

    #[test]
    fn set_multi_value_option_emits_flag_once() {
        // No mtui registry arg is a non-Append multi-value option, so pin the
        // branch on a bespoke parser (as `store_false_off_side_emits_flag`
        // does): `num_args(2..)` consumes many values from ONE occurrence, so
        // the flag must be emitted once followed by all items.
        let parser = clap::Command::new("probe").no_binary_name(true).arg(
            clap::Arg::new("point")
                .long("point")
                .action(ArgAction::Set)
                .num_args(2..),
        );
        let out = kwargs_to_argv(
            &parser,
            json!({ "point": ["1", "2", "3"] }).as_object().unwrap(),
            &[],
        );
        assert_eq!(out, vec!["--point", "1", "2", "3"]);
        let parsed = parser.try_get_matches_from(&out).unwrap();
        let values: Vec<&String> = parsed.get_many::<String>("point").unwrap().collect();
        assert_eq!(values, ["1", "2", "3"]);
    }

    // ------------------------------------------------------------ omission

    #[test]
    fn absent_optional_positional_emits_nothing() {
        // `export filename` (optional positional) omitted → empty argv.
        let out = argv("export", json!({}));
        assert!(out.is_empty(), "absent arg emits nothing: {out:?}");
    }

    #[test]
    fn null_value_is_skipped() {
        let out = argv("openqa_overview", json!({ "days": null, "export": null }));
        assert!(out.is_empty(), "null values skipped: {out:?}");
    }

    // -------------------------------------------------------- base flags

    #[test]
    fn base_template_flag_round_trips() {
        let out = argv("whoami", json!({ "template": "SUSE:Maintenance:1:1" }));
        assert_eq!(out, vec!["--template", "SUSE:Maintenance:1:1"]);
        assert_reparses("whoami", &out);
    }

    #[test]
    fn all_templates_bool_round_trips() {
        let out = argv("whoami", json!({ "all_templates": true }));
        assert_eq!(out, vec!["--all-templates"]);
        assert_reparses("whoami", &out);
    }

    #[test]
    fn per_command_flags_precede_tail_routed_multi() {
        // A plain scalar flag comes before an append flag routed to the tail.
        let out = argv("commit", json!({ "msg": ["m"], "all_templates": true }));
        assert_eq!(out, vec!["--all-templates", "--msg", "m"]);
        assert_reparses("commit", &out);
    }

    // ------------------------------------------------------- mutex group

    #[test]
    fn mutex_group_member_emits_its_own_flag() {
        // `load_template` -a/-k are distinct clap ids; `auto` maps straight to
        // its long flag with no synthetic routing.
        let out = argv("load_template", json!({ "auto": "SUSE:Maintenance:1:1" }));
        assert_eq!(out, vec!["--auto-review-id", "SUSE:Maintenance:1:1"]);
        assert_reparses("load_template", &out);
    }

    // ---------------------------------------------------------- prefix

    #[test]
    fn argv_prefix_is_prepended() {
        // The MCP tool synthesiser fans `config` out per-subcommand and passes
        // the *subparser* here (its args live on `set`, not the parent).
        let parent = parser_for("config");
        let set = parent
            .get_subcommands()
            .find(|c| c.get_name() == "set")
            .expect("config has a `set` subcommand");
        let out = kwargs_to_argv(
            set,
            json!({ "attribute": "template_dir", "value": "/tmp" })
                .as_object()
                .unwrap(),
            &["set".to_owned()],
        );
        assert_eq!(out, vec!["set", "template_dir", "/tmp"]);
        // The full `config set …` argv re-parses through the parent parser.
        assert_reparses("config", &out);
    }

    // --------------------------------------------------------- helpers

    #[test]
    fn help_and_version_ids_are_ignored() {
        let out = argv("whoami", json!({ "help": true, "version": true }));
        assert!(out.is_empty(), "help/version never round-trip: {out:?}");
    }
}
